### PR TITLE
feat(cvsb-16024): expiry date refactor

### PR DESCRIPTION
**Pull request by m-ujaffer**
## Expiry Date Refactor
Expiry date functionality updated to implement a more modular and testable code.
[link to ticket number](https://jira.dvsacloud.uk/browse/CVSB-16024)

## Checklist
- [X] Code has been tested manually
- [X] PR title includes the JIRA ticket number
- [X] Branch is rebased against the latest develop
- [X] Squashed commit contains the JIRA ticket number

### CODE REVIEW

File: src/handlers/expiry/providers/DateProvider.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+  public static getLastDayOfMonthInNextYearISOString(inputDate: Date | string): string {
+    return moment(inputDate)
+      .add(1, "year")
+      .endOf("month")
+      .startOf("day")
+      .toISOString();
+  }
+
+  public static getStartOfMonthInNextYearISOString(inputDate: Date | string): string {
```
Comment (castancu) [inline src/handlers/expiry/providers/DateProvider.ts:132]: some methods that are not being used, can we remove them?

**Review by castancu — COMMENTED** [review]
Amazing work! Some small questions and suggestions, nothing major.

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        console.log(error);
+        error = new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR);
+      }
+      throw error;
+    }
+  }
+
+  public async getMostRecentExpiryDate(systemNumber: string): Promise<Date> {
+    let maxDate = new Date(1970, 1, 1);
+    const testResults = await this.getTestHistory(systemNumber);
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/handlers/expiry/providers/TestDataProvider.ts:62]: Throwing error makes more sense in my opinion than to return a default date. This would make sure that we don't create an expiry using a wrong value for recent expiry. @alex-deloitte  @JonLazariniDD  I would like your feedback on this. Does Dynamo throw an error if no result is found?

File: src/handlers/expiry/providers/DateProvider.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    return moment(inputDate).isSameOrBefore(moment(compareDate).subtract(2, "months").startOf("day"), "days");
+  }
+
+  public static GetEpoc(): Date {
+    return new Date(0);
+  }
+
+
+
+  public static addOneYear(inputDate: Date | string): Date {
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/handlers/expiry/providers/DateProvider.ts:85]: The reason to have separate methods is some are used to create date values e.g. `regnAnniversary` in case of `HgvTrlAnnualTestStrategy`. All comparison date object and ISOString is encapsulated in DateProvider. This adds separation of concern as a result I never need to add moment as a reference in other strategy classes.

File: src/handlers/VehicleTestController.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  /**
+   * A factory method used to fetch a strategy for calculating expiry date based on a strategy mapping JSON file.
+   * @param testTypeForExpiry an input object which is used to calculate expiry based on test type.
+   */
+  public getExpiryStrategy(testTestForExpiry: TestTypeForExpiry): IExpiryDateStrategy {
+    const selectedStrategy = ExpiryDateStrategyFactory.GetExpiryStrategy(testTestForExpiry, this.dateProvider);
+    return selectedStrategy;
+  }
+  public calculateExpiry(testTestForExpiry: TestTypeForExpiry): string {
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/handlers/VehicleTestController.ts:25]: done

File: src/models/ITestResultData.ts [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import { ITestResult } from "./ITestResult";
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/models/ITestResultData.ts:1]: done

**Comment by m-ujaffer** [comment]
Following comments in regards to unit testing were given by Mike Jones:
**Unit Test Scenarios**

     ExpiryDateStrategyFactory 
       - No Config
       - Blank Config
       - Multiple Strategies defined for the same test code
       - No strategy defined for aa test code

    Test Data Provider
        - GetTestHistory 
            - Passed a blank system number
            - Passed as system number that has no tests
            - Bubble up of Valid errors the AWS SDK may throw

        - getMostRecentExpiryDate	
            - No codes that are used for expiry
	    - No submitted tests (but does have tests)
	    - Is passed blank and garbled (invalid) dates

**Unit Test Scenarios for strategies**
- No prior test date, and/or no regn date.
- very old regn date (say 1900 or something where a migrated value might have been migrated in)
- very old test date
- test all the below around the boundaries of months - so sit them either side of an end of month to verify we are managing our calendar manipulation correctly.
- regn 1 year before test date
- regn 1 year and a day before test date
- regn 14 months before test date
- regn 14 months and a day before test date

**Comment by m-ujaffer** [comment]
> egards to unit testing were given by Mike Jones:
> **Unit Test Scenarios**
    - No Config -> build fails 
   - Multiple Strategies defined for the same test code -> User Input error cannot be tested via unit testing as logic would execute based on data provided. This requires test cases by QA.
   - No strategy defined for aa test code -> Same as above

File: src/handlers/expiry/providers/DateProvider.ts (hunk 1 above)
Comment (m-ujaffer) [reply src/handlers/expiry/providers/DateProvider.ts]: removed

File: src/handlers/VehicleTestController.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import { TestDataProvider } from "./expiry/providers/TestDataProvider";
+import { DateProvider } from "./expiry/providers/DateProvider";
+
+@Service()
+export class VehicleTestController implements IVehicleTestController {
+
+  constructor(public dataProvider: TestDataProvider, public dateProvider: DateProvider) {}
+
+  /**
+   * A factory method used to fetch a strategy for calculating expiry date based on a strategy mapping JSON file.
```
Comment (JonLazariniDD) [inline src/handlers/VehicleTestController.ts:18]: 👍

File: src/handlers/expiry/providers/TestDataProvider.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alex-deloitte) [reply src/handlers/expiry/providers/TestDataProvider.ts:62]: I didn't test this behaviour in Dynamodb but I'd rather not return a default date. We will not need another method to check if the testHistory returned a valid date or not

File: src/handlers/expiry/ExpiryDateStrategyFactory.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+
+export interface ExpiryConfig {
+  isHistoryRequired: boolean;
+  isRegistrationRequired: boolean;
+  testTypes: string[];
+  strategy: string;
+}
+
+// TODO: add unit tests for Factory
```
Comment (JonLazariniDD) [inline src/handlers/expiry/ExpiryDateStrategyFactory.ts:22]: I think we did add unit tests for that so we can probably remove it

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            testEndTimestamp: string | number | Date;
+          }) => {
+            return (
+              moment(testResult.testStartTimestamp).isAfter(fromDateTime) &&
+              moment(testResult.testEndTimestamp).isBefore(toDateTime)
+            );
+          }
+        );
+        result = result.filter(
+          (test) => test.testStatus === TEST_STATUS.SUBMITTED
```
[the remark this answers is not in the document]
Comment (JonLazariniDD) [reply src/handlers/expiry/providers/TestDataProvider.ts:47]: and could potentially also chain the filter operations as this returns a new array I think.
Let me know what you think please?

```
 if (data && data.Count) return result
  .filter(({testStatus}) => testStatus === TEST_STATUS.SUBMITTED)
  .filter(({ testStartTimestamp, testEndTimestamp}) => 
    moment(testStartTimestamp).isAfter(fromDateTime) && moment(testStartTimestamp).isBefore(toDateTime)
  )

```

File: src/handlers/expiry/providers/TestDataProvider.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (JonLazariniDD) [reply src/handlers/expiry/providers/TestDataProvider.ts:62]: This is a good point and thanks for pointing this out, I think it is important to avoid mutating the original input where possible so we can 'guarantee' that possible side effects.
If my understanding is no wrong we set the date in the date provider which is great and act as the main conduit for it.
I think we could set up a session with the business to agree on how we should handle that, in theory if the Data layer fails we could we still might want to know what should return as an error. Apologies on not helping too much on this but I think it is more checking with architects and the business  requirements we currently have.

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+          ) {
+            filteredTestTypeDates.push(moment(testType.testExpiryDate));
+          }
+        });
+      });
+      if (filteredTestTypeDates && filteredTestTypeDates.length > 0) {
+        maxDate = moment.max(filteredTestTypeDates).toDate();
+      }
+    }
+    return maxDate;
```
Comment (JonLazariniDD) [inline src/handlers/expiry/providers/TestDataProvider.ts:84]: Is it worth adding a few more when  `testExpiryDate` are valid so we can see the output of the function in the unit test with the maxDate?

File: src/handlers/expiry/strategies/HgvTrlAnnualTestStrategy.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { DateProvider } from "../providers/DateProvider";
+import { IExpiryDateStrategy } from "../IExpiryDateStrategy";
+import { TestTypeForExpiry } from "../../../models/TestTypeforExpiry";
+
+// TODO:
```
Comment (JonLazariniDD) [inline src/handlers/expiry/strategies/HgvTrlAnnualTestStrategy.ts:5]: i think this is covered we can remove it 👍
Nice one

File: tests/unit/expiry/TestDataProvider.unitTest.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+                });
+                testDataProvider.testResultsDAO = new MockTestResultsDAO();
+                await testDataProvider.getTestHistory("123");
+            } catch (error) {
+                expect.assertions(1);
+                expect(error).toEqual(new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR));
+            }
+            });
+        });
+        describe("when getBySystemNumber returns invalid testEndTimestamp", () => {
```
Comment (JonLazariniDD) [inline tests/unit/expiry/TestDataProvider.unitTest.ts:31]: indentation is weird, we might want to add prettier as an improvement ticket I think

**Comment by JonLazariniDD** [comment]
This is a very nice piece of work 👏

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  TEST_CODES_FOR_CALCULATING_EXPIRY
+} from "../../../assets/Enums";
+
+import { ITestResult } from "../../../models/ITestResult";
+import { TestResultsDAO } from "../../../models/TestResultsDAO";
+import { DateProvider } from "./DateProvider";
+import { ITestDataProvider } from "./ITestDataProvider";
+import { Service } from "../../../models/injector/ServiceDecorator";
+
+// TODO: Add unit tests data provider
```
Comment (JonLazariniDD) [inline src/handlers/expiry/providers/TestDataProvider.ts:15]: I think this also covered isn't ? 👍

File: src/models/injector/GenericClassDecorator.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Type for what object is instances of
+ */
+export type Type<T> = new(...args: any[]) => T;
```
Comment (JonLazariniDD) [inline src/models/injector/GenericClassDecorator.ts:4]: very smart, nice utility 👍

File: src/models/injector/Injector.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
+import "reflect-metadata";
+import {Type} from "./GenericClassDecorator";
+
+/**
+ * The Injector stores services and resolves requested instances.
+ */
+export const Injector = new class {
```
Comment (JonLazariniDD) [inline src/models/injector/Injector.ts:7]: Nice way to deal with dependency injections without having to add another framework 👍

File: src/services/TestResultsService.ts [hunk 15]
```diff
@@ -62,16 +63,22 @@ import {
   testTypesSchemaSpecTestGroup1, testTypesSchemaSpecTestGroup2,
   testTypesSchemaSpecTestGroup3, testTypesSchemaSpecTestGroup4, testTypesSchemaSpecTestGroup5
 } from "../models/test-types/testTypesSchemaSpecialistTestsPut";
+import { Injector } from "../models/injector/Injector";
```
Comment (JonLazariniDD) [inline src/services/TestResultsService.ts:66]: image

👍

File: src/services/TestResultsService.ts [hunk 16]
```diff
@@ -28,7 +27,8 @@ import {
   TEST_TYPES_GROUP4_SPEC_TEST,
   TEST_TYPES_GROUP3_SPEC_TEST,
   TEST_TYPES_GROUP2_SPEC_TEST,
-  SPECIALIST_TEST_TYPE_IDS
+  SPECIALIST_TEST_TYPE_IDS,
```
Comment (JonLazariniDD) [inline src/services/TestResultsService.ts:30]: could probably load those as named imports to reduce the number of lines we have as imports.

File: src/services/TestResultsService.ts [hunk 17]
```diff
@@ -62,16 +63,22 @@ import {
   testTypesSchemaSpecTestGroup1, testTypesSchemaSpecTestGroup2,
   testTypesSchemaSpecTestGroup3, testTypesSchemaSpecTestGroup4, testTypesSchemaSpecTestGroup5
 } from "../models/test-types/testTypesSchemaSpecialistTestsPut";
+import { Injector } from "../models/injector/Injector";
+import { TestDataProvider } from "../handlers/expiry/providers/TestDataProvider";
+import { TestTypeForExpiry } from "../models/TestTypeforExpiry";
```
Comment (JonLazariniDD) [inline src/services/TestResultsService.ts:68]: Could doing the same for the models or export * from an index.ts in this folder so we have an isolated module.
Up to you but it could shorten the imports even more

File: src/services/TestResultsService.ts [hunk 18]
```diff
@@ -62,16 +63,22 @@ import {
...
  * @returns Promise
  */
 export class TestResultsService {
   public readonly testResultsDAO: TestResultsDAO;
-
+  public vehicleTestController: VehicleTestController;
   constructor(testResultsDAO: TestResultsDAO) {
     this.testResultsDAO = testResultsDAO;
+    this.vehicleTestController = Injector.resolve<VehicleTestController>(VehicleTestController, [TestDataProvider, DateProvider]);
+    this.vehicleTestController.dataProvider.testResultsDAO = this.testResultsDAO;
```
Comment (JonLazariniDD) [inline src/services/TestResultsService.ts:81]: I think we flip the !filters to throw we might get away with one nesting condition if it worked

Comment (JonLazariniDD) [reply src/services/TestResultsService.ts:81]: If we use optional chaining we could also remove more nesting. Could be out of scope so happy to discuss and get your take on that please
`filters?.fromDateTime` etc..

File: src/services/TestResultsService.ts [hunk 19]
```diff
@@ -579,121 +586,56 @@ export class TestResultsService {
...
+    try {
+      if (payload.testStatus !== TEST_STATUS.SUBMITTED ||
+          this.isNotAllowedVehicleTypeForExpiry(payload.vehicleType)) {
+        return Promise.resolve(payload);
+      }
+      const expiryTestTypes = payload.testTypes.filter((testType) => this.isAllowedTestTypeForExpiry(testType));
+
+      const recentExpiry =  await this.vehicleTestController.dataProvider.getMostRecentExpiryDate(payload.systemNumber);
+
+      expiryTestTypes.forEach((testType: any, index: number) => {
```
Comment (JonLazariniDD) [inline src/services/TestResultsService.ts:600]: up to you but we could be chaining the forEach() after the filter.
I will have a question for this function, so I would love to have a chat about that whenever you are available please.

File: src/models/injector/GenericClassDecorator.ts (hunk 13 above)
Comment (m-ujaffer) [reply src/models/injector/GenericClassDecorator.ts:4]: This is actually used in all of our repos for DVSA. Nothing special from my end :)

File: src/handlers/expiry/providers/TestDataProvider.ts [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+  public async getMostRecentExpiryDate(systemNumber: string): Promise<Date> {
+    let maxDate = DateProvider.getEpoc();
+    const testResults = await this.getTestHistory(systemNumber);
+    if (testResults && testResults.length) {
+      const filteredTestTypeDates: any[] = [];
+      testResults.forEach((testResult) => {
+        testResult.testTypes.forEach((testType) => {
+          if (
+            testType.testCode &&
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply src/handlers/expiry/providers/TestDataProvider.ts:68]: updated the code to destructure props, the reduce doesn't work.

File: src/handlers/expiry/ExpiryDateStrategyFactory.ts (hunk 7 above)
Comment (m-ujaffer) [reply src/handlers/expiry/ExpiryDateStrategyFactory.ts:22]: removed comments

File: src/handlers/expiry/providers/TestDataProvider.ts (hunk 9 above)
Comment (m-ujaffer) [reply src/handlers/expiry/providers/TestDataProvider.ts:84]: Added one scenario with all valid dates and added one scenario with an invalid date.

File: src/handlers/expiry/strategies/HgvTrlAnnualTestStrategy.ts (hunk 10 above)
Comment (m-ujaffer) [reply src/handlers/expiry/strategies/HgvTrlAnnualTestStrategy.ts:5]: removed comments

File: tests/unit/expiry/TestDataProvider.unitTest.ts (hunk 11 above)
Comment (m-ujaffer) [reply tests/unit/expiry/TestDataProvider.unitTest.ts:31]: fixed indentation for all unit tests.

File: src/handlers/expiry/providers/TestDataProvider.ts (hunk 12 above)
Comment (m-ujaffer) [reply src/handlers/expiry/providers/TestDataProvider.ts]: removed comments

**Review by alex-deloitte — APPROVED** [review]

**Review by JonLazariniDD — APPROVED** [review]

File: tests/unit/expiry/HgvTrlMostRecentExpiryStrategy.unitTest.ts [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+  });
+
+  afterEach(() => {
+    // reset date to current date
+    hgvTrlMostRecentExpiryStrategy.dateProvider.setTestDate(new Date());
+  });
+
+  context("for hgv vehicle type", () => {
+    describe("with missing expiry date", () => {
+      it("should set the Expiry last day of month + 1 year of testDate", () => {
```
Comment (bristol-mike-jones) [inline tests/unit/expiry/HgvTrlMostRecentExpiryStrategy.unitTest.ts:24]: We should be testing more than the last day of the month.  We should check that the first day of the month, middle.  We should check that on leap years we wind forward to the 29th.  If this used a test.each table - we could quickly and easily test the start, middle and end of every month in a few dozen lines.

The recent expiry is passed as the epoch date - shouldn't this be empty/undefined as that is what the rule is.  Also the regn date is passed, we should test for this being present and absent (think the rules are the same)

File: tests/unit/expiry/HgvTrlMostRecentExpiryStrategy.unitTest.ts [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+          new Date("2020-06-30")
+        );
+        const expectedExpiryDate = new Date("2021-06-30").toISOString();
+        const actualExpiryDate = hgvTrlMostRecentExpiryStrategy.getExpiryDate();
+        expect.assertions(1);
+        expect(actualExpiryDate).toEqual(expectedExpiryDate);
+      });
+    });
+    describe("with valid expiry date", () => {
+      describe("which is exactly 2 months of test date", () => {
```
Comment (bristol-mike-jones) [inline tests/unit/expiry/HgvTrlMostRecentExpiryStrategy.unitTest.ts:55]: Again, we should be testing more than a single value for the condition, and for the calculation.  We should be checking the two-month window on boundaries of months, including leap years. and the calculation should check that we add a year to a leap year correctly.

File: tests/unit/expiry/HgvTrlFirstTestStrategy.unitTest.ts [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+  afterEach(() => {
+    // reset date to current date
+    hgvTrlFirstTestStrategy.dateProvider.setTestDate(new Date());
+  });
+
+  context("for hgv vehicle type", () => {
+    describe("with a valid registration date", () => {
+      describe(" and the test is conducted exactly 2 month from registration anniversary", () => {
```
Comment (bristol-mike-jones) [inline tests/unit/expiry/HgvTrlFirstTestStrategy.unitTest.ts:24]: There are 4 scenarios given on https://wiki.dvsacloud.uk/display/HVT/Expiry+Date+Logic+-+HGV  but only a couple of unit tests here.  Where are the others? Also, we should be testing boundaries and leap years etc. rather than a single case.

Using the jest tables to input the scenario's would make this much easier to read, make the test code more concise (it's very WET) as well as making it easy to get the business to sign off on the test cases

**Review by bristol-mike-jones — CHANGES_REQUESTED** [review]

File: tests/unit/expiry/ExpiryDateStrategyFactory.unitTest.ts [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+            new DateProvider()
+          ).getExpiryDate();
+        } catch (error) {
+          expect.assertions(1);
+          expect(error).toEqual(new Error(ERRORS.MethodNotImplemented));
+        }
+      });
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply tests/unit/expiry/ExpiryDateStrategyFactory.unitTest.ts:71]: Define error, as I have already added a test which has 
- invalid vehicle type, 
- test with a missing json. 
An invalid json will fail in compilation which consequently causes the build to fail. 

I need to first identify a scenario which doesn't match a strategy TestTypeForExpiry. I will update accordingly.

In my opinion handling for returning a more than one strategy is an improvement in the code as the current implementation is implemented with a valid strategy. However, I can adjust this implementation.

File: tests/unit/expiry/PsvMostRecentExpiryStrategy.unitTest.ts [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+          new Date("2020-05-29")
+        );
+        const expectedExpiryDate = new Date("2021-07-28").toISOString();
+        const actualExpiryDate = psvMostRecentExpiryStrategy.getExpiryDate();
+        expect.assertions(1);
+        expect(actualExpiryDate).toEqual(expectedExpiryDate);
+      });
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply tests/unit/expiry/PsvMostRecentExpiryStrategy.unitTest.ts:46]: unit tests updated according to the suggested format.

File: tests/unit/expiry/HgvTrlAnnualTestStrategy.unitTest.ts [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+  beforeEach(() => {
+    testResultsMockDB = cloneDeep(testResults);
+  });
+
+  afterEach(() => {
+    // reset date to current date
+    hgvTrlAnnualTestStrategy.dateProvider.setTestDate(new Date());
+  });
+
+  context("for hgv vehicle type", () => {
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply tests/unit/expiry/HgvTrlAnnualTestStrategy.unitTest.ts:22]: All tests replaced with test table format.

File: tests/unit/expiry/HgvTrlAnnualTestStrategy.unitTest.ts [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+          new Date("2020-06-30")
+        );
+        const expectedExpiryDate = new Date("2021-06-30").toISOString();
+        const actualExpiryDate = hgvTrlAnnualTestStrategy.getExpiryDate();
+        expect.assertions(1);
+        expect(actualExpiryDate).toEqual(expectedExpiryDate);
+      });
+    });
+    describe("with valid registration date", () => {
+      describe("with a registration anniversary which is 1 before 2 months of test date", () => {
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply tests/unit/expiry/HgvTrlAnnualTestStrategy.unitTest.ts]: all existing tests updated according to the test table format.

File: tests/unit/expiry/HgvTrlAnnualTestStrategy.unitTest.ts [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+          hgvTrlAnnualTestStrategy.dateProvider.setTestDate(
+            new Date("2021-01-01")
+          );
+          const expectedExpiryDate = new Date("2022-02-28").toISOString();
+          const actualExpiryDate = hgvTrlAnnualTestStrategy.getExpiryDate();
+          expect.assertions(1);
+          expect(actualExpiryDate).toEqual(expectedExpiryDate);
+        });
+      });
+      describe("with registration anniversary same  as test date", () => {
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply tests/unit/expiry/HgvTrlAnnualTestStrategy.unitTest.ts]: All existing tests updated according to the test table format.

File: tests/unit/expiry/PsvRegistrationAnniversaryStrategy.unitTest.ts [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+          testTypeForExpiry,
+          new DateProvider()
+        );
+        psvRegistrationAnniversaryStrategy.dateProvider.setTestDate(
+          new Date("2020-05-28")
+        );
+        const expectedExpiryDate = new Date("2021-05-27").toISOString();
+        const actualExpiryDate = psvRegistrationAnniversaryStrategy.getExpiryDate();
+        expect.assertions(1);
+        expect(actualExpiryDate).toEqual(expectedExpiryDate);
```
[the remark this answers is not in the document]
Comment (m-ujaffer) [reply tests/unit/expiry/PsvRegistrationAnniversaryStrategy.unitTest.ts]: All existing tests updated according to the jest table format.

File: tests/unit/expiry/HgvTrlMostRecentExpiryStrategy.unitTest.ts (hunk 21 above)
Comment (m-ujaffer) [reply tests/unit/expiry/HgvTrlMostRecentExpiryStrategy.unitTest.ts]: The tests are replaced with jest table approach and the test for recent Expiry to be epoch or undefined is invalid in regards to `HgvTrlMostRecentExpiryStrategy` as this strategy requires a valid recent expiry. This test case is removed.

File: tests/unit/expiry/HgvTrlFirstTestStrategy.unitTest.ts (hunk 23 above)
Comment (m-ujaffer) [reply tests/unit/expiry/HgvTrlFirstTestStrategy.unitTest.ts]: All existing tests updated according to the jest table format and strategy is mocked to make the code DRY.

**Comment by bristol-mike-jones** [comment]
Hi Umer - This is much clearer thank you, Great work!
